File: source/graphics/vertex_attribute.cpp
#include "../../include/graphics/vertex_attribute.h"

#include <vector>
#include <glm/glm.hpp>

VertexAttribute VertexAttribute::Int32 = VertexAttribute::create(GL_INT, sizeof(int32_t), 1);
VertexAttribute VertexAttribute::Int64 = VertexAttribute::create(GL_INT, sizeof(int64_t), 1);
VertexAttribute VertexAttribute::Float = VertexAttribute::create(GL_FLOAT, sizeof(float), 1);
VertexAttribute VertexAttribute::Vector2 = VertexAttribute::create(GL_FLOAT, sizeof(glm::vec2), 2);
VertexAttribute VertexAttribute::Vector3 = VertexAttribute::create(GL_FLOAT, sizeof(glm::vec3), 3);
VertexAttribute VertexAttribute::Vector4 = VertexAttribute::create(GL_FLOAT, sizeof(glm::vec4), 4);

VertexAttribute::VertexAttribute(GLenum pointerType, unsigned int sizeInBytes, unsigned int componentsCount) {
	this->pointerType = pointerType;
	this->sizeInBytes = sizeInBytes;
	this->componentsCount = componentsCount;
}
VertexAttribute VertexAttribute::create(GLenum pointerType, unsigned int sizeInBytes, unsigned int componentsCount) {
	return *new VertexAttribute(pointerType, sizeInBytes, componentsCount);
}
unsigned int VertexAttribute::get_total_size_in_bytes(std::vector<VertexAttribute> attributes) {
	unsigned int size = 0;
	for (int i = 0; i < attributes.size(); ++i)
	{
		size += attributes[i].sizeInBytes;
	}
	return size;
}

File: CSharp/ASE_CSharp/Graphics/VertexAttribute.cs
using OpenTK.Graphics.OpenGL4;

namespace ASE.Graphics
{
    public class VertexAttribute
    {
        public int sizeInBytes;
        public int componentsCount;
        public VertexAttribPointerType pointerType;

        public static VertexAttribute Int16 = new VertexAttribute(VertexAttribPointerType.Int, sizeof(Int16), 1);
        public static VertexAttribute Int32 = new VertexAttribute(VertexAttribPointerType.Int, sizeof(Int32), 1);
        public static VertexAttribute Int64 = new VertexAttribute(VertexAttribPointerType.Int, sizeof(Int64), 1);
        public static VertexAttribute Float = new VertexAttribute(VertexAttribPointerType.Float, sizeof(Int32), 1);
        public static VertexAttribute Vector2 = new VertexAttribute(VertexAttribPointerType.Float, sizeof(float) * 2, 2);
        public static VertexAttribute Vector3 = new VertexAttribute(VertexAttribPointerType.Float, sizeof(float) * 3, 3);
        public static VertexAttribute Vector4 = new VertexAttribute(VertexAttribPointerType.Float, sizeof(float) * 4, 4);

        public VertexAttribute(VertexAttribPointerType pointerType, int sizeInBytes, int componentsCount)
        {
            this.pointerType = pointerType;
            this.sizeInBytes = sizeInBytes;
            this.componentsCount = componentsCount;
        }
        public static int GetTotalSizeInBytes(params VertexAttribute[] attributes)
        {
            int size = 0;
            for (int i = 0; i < attributes.Length; i++)
            {
                size += attributes[i].sizeInBytes;
            }
            return size;
        }
    }
}
